tools/splat_ext/pm_sbn.py

```diff
@@ -0,0 +1,884 @@
+
+            entry = BufferEntry(file_id, bank_index, bank_group)
+
+            # Replace self.init.bk_entries[id]
+            if id < len(self.init.bk_entries):
+                print(f"Overwriting bank ID {id:02X}")
+                self.init.bk_entries[id] = entry
+            elif id == len(self.init.bk_entries):
+                self.init.bk_entries.append(entry)
+            else:
+                raise Exception(f"Invalid bank ID: {id:02X} - cannot have gaps")
+
+        return [config_file_path]
+
+
+    def lookup_file_id(self, filename: str) -> int:
+        for id, sbn_file in enumerate(self.files):
+            if sbn_file.file_name() == filename:
+                return id
+        raise Exception(f"File not found: {filename} - is it in the file_id_map?")
+
+    def __str__(self) -> str:
+        s = "SBN(\n"
+        s += "  files=[\n"
+        for sbn_file in self.files:
+            s += f"    {sbn_file},\n"
+        s += "  ],\n"
+        s += f"  init={self.init}\n"
+        s += ")"
+        return s
+
+    def __hash__(self) -> int:
+        h = 0
+        for sbn_file in self.files:
+            h ^= hash(sbn_file)
+        h ^= hash(self.init)
+        return h
+
+@dataclass
+class SBNHeader:
+    signature: bytes = b"SBN "
+    size: int = 0
+    unk_08: int = 0
+    unk_0C: int = 0
+    tableOffset: int = 0
+    numEntries: int = 0
+    unk_18: int = 0x07c0 # TODO figure out what these are, don't hardcode
+    unk_1C: int = 0x07a0 # TODO figure out what these are, don't hardcode
+    unk_20: int = 0
+    INIToffset: int = 0
+    unk_28: int = 0
+    unk_2C: int = 0
+    unk_30: int = 0
+    unk_34: int = 0
+    unk_38: int = 0
+    unk_3C: int = 0
+
+    fstring = ">4siii iiii iiii iiii"
+    length = 0x40
+
+    def __iter__(self):
+        return iter(astuple(self))
+
+    def __post_init__(self):
+        assert(self.signature == b"SBN ")
+
+    
+@dataclass
+class SBNFileEntry:
+    offset: int
+    fmt: int
+    unk__05: int = 0
+    size: int = 0
+    
+    fstring = ">iBBh"
+    length = 0x08
+
+    def __iter__(self):
+        return iter(astuple(self))
+
+class SBNFile:
+    signature: bytes
+    size: int
+
+    name: bytes
+    data: bytes
+
+    host_system_name: str
+    # the "size" field of the SBN File entry usually, but not always, matches the actual file size.
+    # since this field is unused (I think), it doesn't actually matter other than for reproducing the original cart
+    # I couldn't figure out the pattern behind these fake sizes (if there even is one), so I'm just hardcoding them.
+    fakesize: int
+    # The file ID of the SBN file - needed to guarantee unique filenames. the filename in the header can't be used because it's not unique
+    ident: int
+    def decode(self, data: bytes, ident: int) -> int:
+        self.signature, = struct.unpack(">4s", data[0x00:0x04])
+        self.size, = struct.unpack(">i", data[0x04:0x08])
+        self.name, = struct.unpack(">4s", data[0x08:0x0C])
```

Comment:
```suggestion
        self.signature, self.size, self.name = struct.unpack(">4si4s", data[0x00:0x0C])
```

Reply:
we should probably even use unpack from for this